### PR TITLE
Print out more information in tests when we get an exception.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide   Created this to try to help track down errors like the one reported in:

https://github.com/dotnet/roslyn/issues/20700

The error doesn't have the original stacks where things went wrong, making them mostly useless (especially without a dump to use to examine the original exception).  So we now create a useful message to examine so that the test output will be more forthcoming.

File: src/EditorFeatures/TestUtilities/Workspaces/TestWorkspace.cs [hunk 1]
```diff
@@ -166,6 +188,32 @@ protected override void Dispose(bool finalize)
...
+            {
+                AddException(exception, result);
+            }
+
+            return result;
+        }
+
+        private static void AddException(Exception exception, List<Exception> result)
+        {
+            if (exception is AggregateException aggregate)
```
Comment (Pilchie) [inline src/EditorFeatures/TestUtilities/Workspaces/TestWorkspace.cs]: You might also consider handling `TargetInvocationException` specially.

File: src/EditorFeatures/TestUtilities/Workspaces/TestWorkspace.cs [hunk 2]
```diff
@@ -142,15 +143,36 @@ protected override void Dispose(bool finalize)
...
+                var messageBuilder = new StringBuilder();
+                messageBuilder.AppendLine(
+$@"{exceptions.Count} exception(s) were thrown during test.
+Note: exceptions may have been thrown by another test running concurrently with
+this test.  This can happen with any tests that share the same ExportProvider.
+Examining individual exception stacks may help reveal the original test and source 
+of the problem.");
+
+                messageBuilder.AppendLine();
+                for (int i = 0; i < exceptions.Count; i++)
```
Comment (tannergooding) [inline src/EditorFeatures/TestUtilities/Workspaces/TestWorkspace.cs]: AggregateException already does roughly this: https://source.dot.net/#System.Private.CoreLib/src/System/AggregateException.cs,434

File: src/EditorFeatures/TestUtilities/Workspaces/TestWorkspace.cs (hunk 1 above)
Comment (tannergooding) [inline src/EditorFeatures/TestUtilities/Workspaces/TestWorkspace.cs]: AggregateException has a built in `Flatten` method: https://source.dot.net/#System.Private.CoreLib/src/System/AggregateException.cs,389

File: src/EditorFeatures/TestUtilities/Workspaces/TestWorkspace.cs [hunk 3]
```diff
@@ -142,15 +143,36 @@ protected override void Dispose(bool finalize)
                 document.CloseTextView();
             }
 
-            var exceptions = ExportProvider.GetExportedValue<TestExtensionErrorHandler>().GetExceptions();
+            var exceptions = Flatten(ExportProvider.GetExportedValue<TestExtensionErrorHandler>().GetExceptions());
```
Comment (tannergooding) [inline src/EditorFeatures/TestUtilities/Workspaces/TestWorkspace.cs]: Why do we need a custom `Flatten` method here, AggregateException already does (or provides this functionality through its own method) this for us.

Comment (CyrusNajmabadi) [inline src/EditorFeatures/TestUtilities/Workspaces/TestWorkspace.cs]: I'm trying to do the inverse.  I'm trying to unwrap all the exceptions into the final list.  i may have many differnet exceptions, some of which are AggregateExceptions or not.  But i could just reuse their helpers here.  So i'll do that.

File: src/EditorFeatures/TestUtilities/Workspaces/TestWorkspace.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/TestUtilities/Workspaces/TestWorkspace.cs]: done.

**Comment by dpoeschl** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by dpoeschl** [comment]
retest windows_release_vs-integration_prtest please

**Pull request merged**